Improve the performance of the `ProductFormula` synthesizers

mrossinek opened the pull request:
<!--
⚠️ If you do not respect this template, your pull request will be closed.
⚠️ Your pull request title should be short detailed and understandable for all.
⚠️ Also, please add a release note file using reno if the change needs to be
  documented in the release notes.
⚠️ If your pull request fixes an open issue, please link to the issue.

- [ ] I have added the tests to cover my changes.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the CONTRIBUTING document.
-->

### Summary

This PR implements the suggestion made in https://github.com/Qiskit/qiskit/pull/12021#discussion_r1528762352.

### Details and comments

I ran a very simple benchmark to showcase the performance improvements of this PR. Below you can see that we get a ~30x performance improvement (note the logarithmic y axis).

![benchmark](https://github.com/Qiskit/qiskit/assets/21973473/cbaab056-1026-4c36-882f-50ad6178b0d6)

<details>
<summary>Here is the code which builds a heavy-hex graph of a given size, constructs an XYZ model Hamiltonian on its connectivity, and builds a LieTrotter and SuzukiTrotter circuit.</summary>

```python
import sys
import time

import numpy as np
import rustworkx as rx
from qiskit import QuantumCircuit, transpile
from qiskit.circuit.library import PauliEvolutionGate
from qiskit.quantum_info import SparsePauliOp
from qiskit.synthesis import LieTrotter, SuzukiTrotter

for size in range(3, 10, 2):
    graph = rx.generators.heavy_hex_graph(size, multigraph=False)

    num_qubits = graph.num_nodes()

    sparse_pauli_list = []
    for edge in graph.edge_list():
        for p in ("XX", "YY", "ZZ"):
            sparse_pauli_list.append((p, [edge[0], edge[1]], 1.0))
    for qubit in range(num_qubits):
        for p in "XYZ":
            sparse_pauli_list.append((p, [qubit], 1.0))

    hamil = SparsePauliOp.from_sparse_list(sparse_pauli_list, num_qubits=num_qubits)

    print("starting LieTrotter with 10 reps", file=sys.stderr)
    t1 = time.time()
    circuit = QuantumCircuit(num_qubits)
    circuit.append(
        PauliEvolutionGate(hamil, time=0.5, synthesis=LieTrotter(reps=10)),
        qargs=circuit.qubits,
    )
    transpiled = transpile(circuit, basis_gates=["rz", "x", "sx", "cx"])
    t2 = time.time()
    lie = t2 - t1
    print(f"took {lie}s", file=sys.stderr)

    print("starting SuzukiTrotter with 10 reps", file=sys.stderr)
    t1 = time.time()
    circuit = QuantumCircuit(num_qubits)
    circuit.append(
        PauliEvolutionGate(hamil, time=0.5, synthesis=SuzukiTrotter(reps=10)),
        qargs=circuit.qubits,
    )
    transpiled = transpile(circuit, basis_gates=["rz", "x", "sx", "cx"])
    t2 = time.time()
    suzuki = t2 - t1
    print(f"took {suzuki}s", file=sys.stderr)
```
</details>

As part of this improvement, I also aligned the implementations of `LieTrotter` and `SuzukiTrotter` which partially diverged in #12021. I also handle the wrapping of the individually time-evolved Pauli terms in a new way, which results in a different (but imo improved) visualization of a circuit. Below you can see the same circuit drawn with Qiskit 1.1.1 and this PR, side by side:

| **Qiskit 1.1.1** | This PR |
|:-:|:-:|
| ![circuit_qiskit_1 1 1](https://github.com/Qiskit/qiskit/assets/21973473/e55f11de-b2b3-4e78-adc8-127474000c25) | ![circuit_branch](https://github.com/Qiskit/qiskit/assets/21973473/f672db79-05b0-46d3-a7e6-886935069d6b) |

<details>
<summary>Here is the code to plot the circuit above</summary>

```python
import numpy as np
from qiskit import QuantumCircuit
from qiskit.circuit.library import PauliEvolutionGate
from qiskit.quantum_info import SparsePauliOp
from qiskit.synthesis import LieTrotter, SuzukiTrotter

num_qubits = 10
edges = [(i, i + 1) for i in range(0, num_qubits - 1, 2)]
edges += [(i, i + 1) for i in range(1, num_qubits - 1, 2)]

sparse_pauli_list = []
for edge in edges:
    for p in ("XX", "YY", "ZZ"):
        sparse_pauli_list.append((p, [edge[0], edge[1]], 1.0))
for qubit in range(num_qubits):
    for p in "XYZ":
        sparse_pauli_list.append((p, [qubit], 1.0))

hamil = SparsePauliOp.from_sparse_list(sparse_pauli_list, num_qubits=num_qubits)

circuit = QuantumCircuit(num_qubits)
circuit.append(
    PauliEvolutionGate(hamil, time=0.5, synthesis=LieTrotter(wrap=True)),
    qargs=circuit.qubits,
)
circuit = circuit.decompose()
circuit.draw("mpl", filename="circuit_branch.png")
```
</details>

mrossinek committed:
[WIP] adds the output argument to the internal atomic evolution

mrossinek committed:
meta: modernize type hints

mrossinek committed:
refactor: change callable structure of atomic evolution

This changes the structure of the `atomic_evolution` callable in the
`ProductFormula` synthesis class. This is motivated by the significant
performance improvements that can be obtained by appending to the
existing circuit directly rather than building out individual evolution
circuits and iteratively composing them.

mrossinek committed:
refactor: deprecate the legacy atomic_evolution signature

mrossinek committed:
refactor: add the wrap argument to ProductFormula

This can be used to recover the previous behavior in which the single
individually evolved Pauli terms get wrapped into gate objects.

mrossinek committed:
fix: insert the missing barriers between LieTrotter repetitions

mrossinek committed:
refactor: align SuzukiTrotter and LieTrotter

mrossinek committed:
Propoagate deprecation notice

mrossinek committed:
fix: the labels of wrapped Pauli evolutions

mrossinek committed:
fix: respect the insert_barriers setting

mrossinek committed:
docs: add a release note

Cryoris commented on qiskit/synthesis/evolution/lie_trotter.py:
```diff
@@ -47,55 +52,70 @@ class LieTrotter(ProductFormula):
+            "'atomic_evolution' argument"
+        ),
+        additional_msg=(
+            "Instead you should update your 'atomic_evolution' function to be of the following "
+            "type: 'Callable[[QuantumCircuit, Pauli | SparsePauliOp, float], None]'."
```
It would be nice to explain a bit more clearly how this should look like, maybe we could just refer to the class docstring and explain in the class docs that `atomic_evolution` should perform the evolution in place? I'll leave a suggestion below.

Cryoris commented on qiskit/synthesis/evolution/lie_trotter.py:
```diff
@@ -47,55 +52,70 @@ class LieTrotter(ProductFormula):
+                ``atomic_evolution is None``.
             atomic_evolution: A function to construct the circuit for the evolution of single
                 Pauli string. Per default, a single Pauli evolution is decomposed in a CX chain
                 and a single qubit Z rotation.
```
Maybe something along these lines?
```suggestion
            atomic_evolution: A function to apply the evolution of a single :class:`.Pauli`,
                or :class:`.SparsePauliOp` of only commuting terms, to a circuit.
                This function takes the circuit to evolve as first argument, followed by
                the Pauli operator evolve, then the evolution time. Per default, a single
                Pauli evolution is decomposed in a CX chain and a single RZ gate.
```

Cryoris commented on releasenotes/notes/product-formula-improvements-1bc40650151cf107.yaml:
```diff
@@ -0,0 +1,29 @@
+    barriers between circuit repetitions.
+features_synthesis:
+  - |
+    Added the ``wrap`` keyword argument to the :class:`~ProductFormula` classes
```
I think this should be (also applies to some other places in the reno)
```suggestion
    Added the ``wrap`` keyword argument to the :class:`.ProductFormula` classes
```
afaik
* the `.` triggers a search that matches with the name that follows (since we don't give the full path we need it here I think)
* the `~` truncates the path to the object, since we don't give the path, we don't need to truncate it (instead we'd need it e.g. if we did :class:`~qiskit.synthesis.evolution.ProductFormula`)

Cryoris commented on releasenotes/notes/product-formula-improvements-1bc40650151cf107.yaml:
```diff
@@ -0,0 +1,29 @@
+deprecations_synthesis:
+  - |
+    The ``atomic_evolution`` argument to :class:`~ProductFormula` (and its
+    subclasses) has deprecated its old function signature.
```
Maybe we could be a bit more explicit, like you have above already
```suggestion
    In :class:`.ProductFormula` and subclasses, it is deprecated that ``atomic_evolution``
    returns a new circuit implementing the evolution. Instead, the evolution must be done
    in place, which allows for a performance benefit.
```

Cryoris commented on qiskit/synthesis/evolution/suzuki_trotter.py:
```diff
@@ -51,15 +55,32 @@ class SuzukiTrotter(ProductFormula):
+            | Callable[[QuantumCircuit, Pauli | SparsePauliOp, float], None]
+            | None
+        ) = None,
+        wrap: bool = False,
```
I think this class misses the forward of `wrap` to the superclass 🤔

Cryoris reviewed (commented):
This looks great, both performance and visualization wise, thanks for the effort! 🙂 I just have a few minor comments below.

mrossinek committed:
Apply suggestions from code review

Co-authored-by: Julien Gacon <gaconju@gmail.com>

mrossinek committed:
fix: missing `wrap` forward in SuzukiTrotter

mrossinek committed:
docs: improve documentation of the `atomic_evolution` argument

Co-authored-by: Julien Gacon <gaconju@gmail.com>

mrossinek committed:
docs: also document the deprecated form of `atomic_evolution`

mrossinek committed:
docs: call out ProductFormula docs in release note

mrossinek committed:
Merge branch 'main' into local_atomic_evolution

mrossinek committed:
Merge branch 'main' into local_atomic_evolution

mrossinek committed:
refactor: change to PendingDeprecationWarning

mrossinek committed:
refactor: explicitly convert to Gate when wrapping

This is slightly faster than the `.compose`-based operation done
previously as it performs fewer checks. Thanks to @jakelishman for the
suggestion offline.

mrossinek commented on qiskit/synthesis/evolution/product_formula.py:
```diff
@@ -58,15 +89,27 @@ def __init__(
+        elif len(inspect.signature(atomic_evolution).parameters) == 2:
+
+            def wrap_atomic_evolution(output, operator, time):
+                definition = atomic_evolution(operator, time)
+                output.compose(definition, wrap=wrap, inplace=True)
```
I was not sure whether we should also do the explicit `Gate` conversion here?
```suggestion
                definition = atomic_evolution(operator, time)
                gate = definition.to_gate()
                output.append(gate, qargs=output.qubits, copy=False)
```
I believe, this will affect the display label. Maybe one can add `label=definition.name` to the `.to_gate()` call to ensure whatever name the user might have set, is preserved? :thinking:

mrossinek replied:
After an offline discussion, we are leaving this with `compose`.

Cryoris commented on qiskit/synthesis/evolution/lie_trotter.py:
```diff
@@ -47,55 +52,76 @@ class LieTrotter(ProductFormula):
+                into a chain of ``CX`` gates and a single ``RZ`` gate.
+                **DEPRECATED:** The function takes only the Pauli operator and evolution time as
+                inputs and returns the circuit that will be appended to the overall circuit being
+                built.
```
Without the deprecation these need updating 🙂 
```suggestion
                Alternatively, the function can also take Pauli operator and evolution time as
                inputs and returns the circuit that will be appended to the overall circuit being
                built.
```

mrossinek committed:
Update qiskit/synthesis/evolution/lie_trotter.py

Co-authored-by: Julien Gacon <gaconju@gmail.com>

mrossinek committed:
docs: update after pending deprecation

Cryoris reviewed (approved):
Great improvement, thanks a lot for the work! 🚤 💨
